File: esw-utils/src/test/scala/esw/highlevel/dsl/EventServiceDslTest.scala
package esw.highlevel.dsl

import csw.event.api.scaladsl.{EventPublisher, EventService, EventSubscriber}
import csw.params.core.models.Prefix
import csw.params.events.{Event, EventKey, EventName, SystemEvent}
import esw.ocs.api.BaseTestSuite
import org.mockito.ArgumentMatchers.{any, eq => argsEq}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.concurrent.duration.DurationDouble

class EventServiceDslImpl(val eventService: EventService) extends EventServiceDsl

class EventServiceDslTest extends BaseTestSuite {
  private val eventService    = mock[EventService]
  private val eventPublisher  = mock[EventPublisher]
  private val eventSubscriber = mock[EventSubscriber]
  private val event           = SystemEvent(Prefix("TCS.test"), EventName("event-1"))

  private val eventServiceDsl = new EventServiceDslImpl(eventService)

  when(eventService.defaultPublisher).thenReturn(eventPublisher)
  when(eventService.defaultSubscriber).thenReturn(eventSubscriber)

  "publish" must {
    "delegate to publishing single event | ESW-120" in {
      eventServiceDsl.publishEvent(event)
      verify(eventService.defaultPublisher).publish(event)
    }

    "delegate to publishing event with generator | ESW-120" in {
      eventServiceDsl.publishEvent(5.seconds)(Some(event))
      verify(eventService.defaultPublisher).publishAsync(
        any[Future[Option[Event]]],
        argsEq(5.seconds)
      )
    }
  }

  "subscribe" must {
    "delegate to subscribing events | ESW-120" in {
      eventServiceDsl.onEvent("TCS.test.event-1")(println)
      verify(eventService.defaultSubscriber).subscribeAsync(any[Set[EventKey]], any[Event => Future[_]]())
    }
  }

  "get" must {
    "delegate to getting events | ESW-120" in {
      eventServiceDsl.getEvent("TCS.test.event-1")
      verify(eventService.defaultSubscriber).get(any[Set[EventKey]])
    }
  }
}

File: script-dsl/src/test/kotlin/esw/ocs/dsl/highlevel/EventServiceDslTest.kt
package esw.ocs.dsl.highlevel

import akka.Done.done
import akka.actor.Cancellable
import csw.event.api.javadsl.IEventPublisher
import csw.event.api.javadsl.IEventSubscriber
import csw.event.api.javadsl.IEventSubscription
import csw.params.core.models.Prefix
import csw.params.events.*
import io.kotlintest.shouldBe
import io.kotlintest.specs.WordSpec
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import java.util.concurrent.CompletableFuture
import kotlin.coroutines.CoroutineContext
import kotlin.time.milliseconds
import kotlinx.coroutines.Job

class EventServiceDslTest : WordSpec({

    class Mocks {
        val key = "TCS.test.eventkey1"
        val duration = 10.milliseconds

        val event: Event = mockk<ObserveEvent>()
        val eventKeys = HashSet<EventKey>()
        val eventSet = HashSet<Event>(1)
        val eventSubscription = mockk<IEventSubscription>()

        val cancellable = mockk<Cancellable>()
        val eventCallback = mockk<(Event) -> CompletableFuture<*>>()

        val eventPublisher: IEventPublisher = mockk()
        val eventSubscriber: IEventSubscriber = mockk()

        init {
            eventKeys.add(EventKey.apply(key))
            eventSet.add(event)
        }

        val eventServiceDsl = object : EventServiceDsl {
            override val coroutineContext: CoroutineContext = Job()
            override val defaultPublisher: IEventPublisher = eventPublisher
            override val defaultSubscriber: IEventSubscriber = eventSubscriber
        }
    }

    "EventServiceDsl" should {

        "systemEvent should return a SystemEvent created with given parameters" {
            with(Mocks()) {
                val eventName = "systemEvent1"
                val eventPrefix = "TCS.filter.wheel"
                val systemEvent = eventServiceDsl.systemEvent(eventPrefix, eventName)

                // Verify that  event with provided prefix and eventName is created.
                systemEvent shouldBe SystemEvent(
                    systemEvent.eventId(),
                    Prefix(eventPrefix),
                    EventName(eventName),
                    systemEvent.eventTime(),
                    systemEvent.paramSet()
                )
            }
        }

        "observeEvent should return a ObserveEvent created with given parameters" {
            with(Mocks()) {
                val eventName = "observeEvent1"
                val eventPrefix = "TCS.filter.wheel"
                val observeEvent = eventServiceDsl.observeEvent(eventPrefix, eventName)

                // Verify that event with provided prefix and eventName is created.
                observeEvent shouldBe ObserveEvent(
                    observeEvent.eventId(),
                    Prefix(eventPrefix),
                    EventName(eventName),
                    observeEvent.eventTime(),
                    observeEvent.paramSet()
                )
            }
        }

        "publish should delegate to publisher.publish" {
            with(Mocks()) {
                every { (eventPublisher.publish(event)) }
                    .returns(CompletableFuture.completedFuture(done()))

                eventServiceDsl.publishEvent(event) shouldBe done()

                verify { eventPublisher.publish(event) }
            }
        }

        "publishEvent should delegate to publisher.publishAsync" {
            with(Mocks()) {
                every { eventPublisher.publishAsync(any(), any()) }.answers { cancellable }

                eventServiceDsl.publishEvent(duration) { event } shouldBe cancellable

                verify { eventPublisher.publishAsync(any(), any()) }
            }
        }

        "onEvent should delegate to subscriber.subscribeAsync" {
            with(Mocks()) {
                every { eventSubscriber.subscribeAsync(eventKeys, any()) }.answers { eventSubscription }

                eventServiceDsl.onEvent(key) { eventCallback } shouldBe eventSubscription

                verify { eventSubscriber.subscribeAsync(eventKeys, any()) }
            }
        }

        "getEvent should delegate to subscriber.get" {
            with(Mocks()) {
                every { eventSubscriber.get(eventKeys) }.answers { CompletableFuture.completedFuture(eventSet) }

                eventServiceDsl.getEvent(key) shouldBe eventSet

                verify { eventSubscriber.get(eventKeys) }
            }
        }
    }
})
